ui/hooks/useSyncEqualityCheck.ts

```diff
@@ -0,0 +1,39 @@
+import { useRef } from 'react';
+import stringify from 'fast-json-stable-stringify';
+
+/**
+ * Synchronous version of `useEqualityCheck` that provides referential stability without triggering unnecessary re-renders.
+ *
+ * Uses JSON serialization to detect changes (including mutations).
+ * Calls `stringify` on every render, `JSON.parse` only when a mutation is detected.
+ * Returns a new reference only when the serialized value differs.
+ * Suitable for small objects/arrays. For large data, consider `useEqualityCheck` or `useMemo`.
+ *
+ * @param value - Must be JSON-serializable. No support for `undefined`, functions, `BigInt`, circular refs, or `Date` objects.
+ * @returns Referentially stable value
+ * @see {@link useEqualityCheck} - async version that triggers re-renders
+ * @example
+ * ```tsx
+ * const stableInput = useSyncEqualityCheck({ chainId, networkClientId });
+ * useEffect(() => {
+ *   startPolling(stableInput);
+ *   return () => stopPolling();
+ * }, [stableInput]);
+ * ```
+ */
+export function useSyncEqualityCheck<Value>(value: Value): Value {
```

Comment:
We can use the `Json` type to constrain this type param to ensure it's JSON-serializable

Reply:
Of course, since we need to be able to pass `value` into `stringify` not just `Object.is`😅

```suggestion
export function useSyncEqualityCheck<Value extends Json>(value: Value): Value {
```